File: src/main/java/com/soukhyatech/faceattendance/service/PiiEncryptionService.java
package com.soukhyatech.faceattendance.service;

import com.soukhyatech.faceattendance.config.JwtConfig;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class PiiEncryptionService {

    private final JwtConfig jwtConfig;
    private static final int GCM_IV_LENGTH = 12;
    private static final int GCM_TAG_LENGTH = 128;
    private static final Set<String> PII_FIELDS = Set.of(
        "aadhaar_number", "pan_number", "phone_no", "email", "card_number"
    );

    private SecretKeySpec keySpec;

    @PostConstruct
    public void init() {
        byte[] keyBytes = jwtConfig.getEncryption().getPiiKey().getBytes(StandardCharsets.UTF_8);
        // Ensure 32 bytes for AES-256
        byte[] key32 = new byte[32];
        System.arraycopy(keyBytes, 0, key32, 0, Math.min(keyBytes.length, 32));
        this.keySpec = new SecretKeySpec(key32, "AES");
    }

    public String encrypt(String plaintext) {
        if (plaintext == null || plaintext.isEmpty()) return null;
        try {
            byte[] iv = new byte[GCM_IV_LENGTH];
            new SecureRandom().nextBytes(iv);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, new GCMParameterSpec(GCM_TAG_LENGTH, iv));
            byte[] encrypted = cipher.doFinal(plaintext.getBytes(StandardCharsets.UTF_8));
            String ivB64 = Base64.getEncoder().encodeToString(iv);
            String cipherB64 = Base64.getEncoder().encodeToString(encrypted);
            return ivB64 + ":" + cipherB64;
        } catch (Exception e) {
            throw new RuntimeException("PII encryption failed", e);
        }
    }

    public String decrypt(String ciphertext) {
        if (ciphertext == null || ciphertext.isEmpty()) return null;
        try {
            String[] parts = ciphertext.split(":");
            byte[] iv = Base64.getDecoder().decode(parts[0]);
            byte[] encrypted = Base64.getDecoder().decode(parts[1]);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.DECRYPT_MODE, keySpec, new GCMParameterSpec(GCM_TAG_LENGTH, iv));
            return new String(cipher.doFinal(encrypted), StandardCharsets.UTF_8);
        } catch (Exception e) {
            return null;
        }
    }

    public String mask(String field, String value) {
        if (value == null || value.isEmpty()) return null;
        return switch (field) {
            case "aadhaar_number" -> "XXXX-XXXX-" + value.substring(value.lastIndexOf('-') + 1);
            case "pan_number" -> "XXXXX" + value.substring(5, 9) + "X";
            case "phone_no" -> value.substring(0, 4) + " *****-" + value.substring(value.length() - 5);
            case "email" -> {
                int at = value.indexOf('@');
                yield value.charAt(0) + "***@" + value.substring(at + 1);
            }
            case "card_number" -> "****" + value.substring(value.length() - 4);
            default -> value;
        };
    }

    public boolean isPiiField(String field) {
        return PII_FIELDS.contains(field);
    }
}

File: csharp/SoukhyaTech.FaceAttendance/Security/PiiEncryptionService.cs
using System.Security.Cryptography;
using System.Text;

namespace SoukhyaTech.FaceAttendance.Security
{
    public class PiiEncryptionService
    {
        private readonly string _key;
        private static readonly HashSet<string> PiiFields = new()
        {
            "aadhaar_number", "pan_number", "phone_no", "email", "card_number"
        };

        public PiiEncryptionService(IConfiguration config)
        {
            _key = config["Encryption:PiiKey"] ?? "dev-pii-key-32-bytes-long!!";
        }

        public string Encrypt(string plaintext)
        {
            if (string.IsNullOrEmpty(plaintext)) return "";
            using var aes = Aes.Create();
            aes.Key = Encoding.UTF8.GetBytes(_key.PadRight(32).Substring(0, 32));
            aes.GenerateIV();
            var iv = aes.IV;
            using var encryptor = aes.CreateEncryptor();
            var bytes = Encoding.UTF8.GetBytes(plaintext);
            var encrypted = encryptor.TransformFinalBlock(bytes, 0, bytes.Length);
            return Convert.ToBase64String(iv) + ":" + Convert.ToBase64String(encrypted);
        }

        public string Decrypt(string ciphertext)
        {
            if (string.IsNullOrEmpty(ciphertext)) return "";
            try
            {
                var parts = ciphertext.Split(':');
                var iv = Convert.FromBase64String(parts[0]);
                var encrypted = Convert.FromBase64String(parts[1]);
                using var aes = Aes.Create();
                aes.Key = Encoding.UTF8.GetBytes(_key.PadRight(32).Substring(0, 32));
                aes.IV = iv;
                using var decryptor = aes.CreateDecryptor();
                var bytes = decryptor.TransformFinalBlock(encrypted, 0, encrypted.Length);
                return Encoding.UTF8.GetString(bytes);
            }
            catch { return ""; }
        }

        public string Mask(string field, string value)
        {
            if (string.IsNullOrEmpty(value)) return "";
            return field switch
            {
                "aadhaar_number" => "XXXX-XXXX-" + value.Substring(value.LastIndexOf('-') + 1),
                "pan_number" => "XXXXX" + value.Substring(5, 4) + "X",
                "phone_no" => value.Substring(0, 4) + " *****-" + value.Substring(value.Length - 5),
                "email" => value[0] + "***@" + value.Split('@')[1],
                "card_number" => "****" + value.Substring(value.Length - 4),
                _ => value
            };
        }

        public bool IsPiiField(string field) => PiiFields.Contains(field);
    }
}
